src/sprite_weapon.cpp

```diff
@@ -0,0 +1,161 @@
+	// Animations for allies
+	// Is a battle charset animation
+
+	int frame = std::min(2, cycle / 10);
+	if (frame == sprite_frame)
+		return;
+
+	const lcf::rpg::BattlerAnimation* anim = lcf::ReaderUtil::GetElement(lcf::Data::battleranimations, battler->GetBattleAnimationId());
+	if (!anim) {
+		Output::Warning("Invalid battler animation ID {}", battler->GetBattleAnimationId());
+		return;
+	}
+
+	const auto* battler_animation_weapon = lcf::ReaderUtil::GetElement(anim->weapons, weapon_animation_id);
+	if (!battler_animation_weapon) {
+		return;
+	}
+
+	SetSrcRect(Rect(frame * 64, battler_animation_weapon->weapon_index * 64, 64, 64));
+
+	const bool flip = battler->IsDirectionFlipped();
+	SetFlipX(flip);
+}
+
+void Sprite_Weapon::SetWeaponAnimation(int nweapon_animation_id) {
+	weapon_animation_id = nweapon_animation_id;
+}
+
+void Sprite_Weapon::StartAttack(bool secondary_weapon) {
+	cycle = 0;
+
+	attacking = true;
+
+	SetZ(default_z + (secondary_weapon ? 1 : -1));
+
+	auto* battler = GetBattler();
+
+	if (battler->GetBattleAnimationId() > 0) {
+		const lcf::rpg::BattlerAnimation* anim = lcf::ReaderUtil::GetElement(lcf::Data::battleranimations, battler->GetBattleAnimationId());
+		if (!anim) {
+			Output::Warning("Invalid battler animation ID {}", battler->GetBattleAnimationId());
+			return;
+		}
+
+		const auto* battler_animation_weapon = lcf::ReaderUtil::GetElement(anim->weapons, weapon_animation_id);
+		if (!battler_animation_weapon) {
+			Output::Warning("Invalid weapon animation ID {}", weapon_animation_id);
+			return;
+		}
+
+		StringView sprite_file = battler_animation_weapon->weapon_name;
+
+		if (!sprite_file.empty()) {
+			FileRequestAsync* request = AsyncHandler::RequestFile("BattleWeapon", sprite_file);
+			request->SetGraphicFile(true);
+			request_id = request->Bind(&Sprite_Weapon::OnBattleWeaponReady, this, battler_animation_weapon->weapon_index);
+			request->Start();
+		}
+	}
+}
+
+void Sprite_Weapon::StopAttack() {
+	attacking = false;
+}
+
+void Sprite_Weapon::CreateSprite() {
+	auto* battler = GetBattler();
+
+	SetX(battler->GetDisplayX());
+	SetY(battler->GetDisplayY());
+
+	SetOx(32);
+	SetOy(32);
+	ResetZ();
+}
+
+void Sprite_Weapon::OnBattleWeaponReady(FileRequestResult* result, int32_t weapon_index) {
+	SetBitmap(Cache::Battleweapon(result->file));
+	SetSrcRect(Rect(0, weapon_index * 64, 64, 64));
+}
+
+void Sprite_Weapon::Draw(Bitmap& dst) {
+	if (!attacking) {
+		return;
+	}
+
+	auto* battler = GetBattler();
+
+	SetTone(Main_Data::game_screen->GetTone());
+	SetX(battler->GetDisplayX());
+	SetY(battler->GetDisplayY());
+	SetFlashEffect(battler->GetFlashColor());
+
+	Sprite_Battler::Draw(dst);
+}
+
+void Sprite_Weapon::ResetZ() {
+	constexpr int id_limit = 128;
+
+	const auto& graphic = GetBitmap();
+	int y = battler->GetBattlePosition().y;
+
+	int z = battler->GetType();
+	z *= SCREEN_TARGET_HEIGHT * 2;
+	z += y;
+	z *= id_limit;
+	z += id_limit - battle_index;
+	z += Priority_Battler;
+
+	default_z = z;
+
+	SetZ(z);
```

Comment:
One suggestion to reduce a bit code duplication:

Instead of ``Game_Actor*`` pass a ``Sprite_Battler*`` (the one that draws the Sprite of the Actor in battle) and store this ``Sprite_Battler`` in a variable.
You can still make the base class constructor happy with ``Sprite_Battler(sprite_actor->GetBattler(), sprite_actor->GetBattler()->GetId())``.

ResetZ can be now replaced with a call to ``sprite_actor->GetZ()`` and the stuff needed for ``CreateSprite`` and ``Draw`` can be also directly read from the ``sprite_actor``

Reply:
I'm probably missing something obvious, but the suggestion does not work for me. Moreover the sprite_battler class has not got the DisplayX, DisplayY and GetFlashColor functions which are used in the Draw function from sprite_weapon.

Reply:
actually you can completely remove ``ResetZ`` because the entire implementation is in ``Sprite_Battler``.

Yes it lacks DisplayX, DisplayY and GetFlashColor but you can use sprite_actor->GetX, sprite_actor->GetY and sprite_actor->GetFlashEffect.

Reply:
The `ResetZ` function has been removed now from `Sprite_Weapon`.